File: python/src/basecamp/_person_id.py
"""Reading a person id off the wire the way Go's ``strconv.ParseInt(s, 10, 64)`` reads one.

BC3 serializes person ids as JSON strings in some responses and as JSON numbers
in others, and it spells the system actors -- ``LocalPerson``, whose id is
``"basecamp"`` or ``"campfire"`` -- in the same string field. Three places in
this SDK have to turn that string into a number: the pre-decode normalizer that
every response body walks (``generated/services/_base.py`` and its async twin),
the flexible id reader in the recording-summary decode
(``services/_campfire_index.py``), and the merge-safe composites' id-list guard
(``services/_merge_safe.py``). The rule lives HERE so they cannot drift apart;
the first two had two copies of it, and the copies disagreed.

So does the WALK that finds those objects (:func:`normalize_person_ids`), for the
same reason: the sync and async base services had a copy each, and which people a
walk finds is as much of the rule as what it does when it finds them.

The grammar is Go's and nothing looser. One optional ASCII ``+`` or ``-``, then
one or more ASCII digits, and nothing else. Python's built-in ``int()`` is the
wrong tool in four separate directions, each of them ACCEPTING -- and every
acceptance here mints an id BC3 never wrote, or collapses a real person onto the
system-actor ``0``:

* ``int()`` strips surrounding whitespace, so ``int(" 7")`` is 7 where Go's scan
  refuses the leading space (Go trims nothing).
* ``int()`` honours PEP 515 underscores, so ``int("1_0")`` is ten. Go allows
  ``_`` only at base 0; at base 10 it is not a digit.
* ``int()`` accepts every Unicode decimal digit, so ``int("７")`` -- fullwidth --
  is 7, and ``int("7৭")`` is 77. Go tests ``'0' <= c && c <= '9'`` on BYTES.
* ``int()`` is arbitrary-precision, so ``"18446744073709551616"`` parses instead
  of overflowing. The wire id is an ``int64``.

Not the same rule as the gid person id
--------------------------------------
``basecamp.mentions.person_id_from_sgid`` reads the id out of
``gid://bc3/Person/<id>`` under a DIFFERENT rule, deliberately: the reference
walks the bytes and refuses anything outside ``0-9`` BEFORE it parses
(``go/pkg/basecamp/mentions.go:252-256``) and then refuses ``id <= 0``
(``:258``), so it rejects a leading ``+`` that this function accepts. That shape
is CORRECT at that site -- relaxing it to this one reintroduces the ``+77``
defect PR #886 closed -- and it is WRONG here, because the Go lines governing
these two sites (``go/pkg/types/flexible_int64.go:34`` and
``go/pkg/basecamp/normalize.go:45``) have no pre-walk at all. Two rules, two
sites, on purpose: do not hoist either into the other, in either direction.
"""

from __future__ import annotations

import json
import re
from collections.abc import Iterable, Mapping, MutableMapping
from enum import Enum
from itertools import islice
from typing import Any
from urllib.parse import urlsplit

from basecamp._security import truncate as _truncate
from basecamp.errors import ApiError

_UINT64_MAX = 2**64 - 1
_INT64_MIN = -(2**63)
_INT64_MAX = 2**63 - 1


class Refusal(Enum):
    """Which way ``strconv.ParseInt`` refused.

    Keeping the two apart is the whole point: Go answers ``0`` to one and an
    error to the other, and no single "is this a number?" predicate can tell
    them apart, because WHICH refusal comes first depends on where in the string
    each disqualifying byte sits (see :func:`parse_int64`).
    """

    SYNTAX = "syntax"
    RANGE = "range"


def parse_int64(text: str) -> int | Refusal:
    """``strconv.ParseInt(text, 10, 64)``, scan order included.

    Returns the value, or the :class:`Refusal` Go would have raised.

    The subtlety worth the hand-rolled loop: ``ParseInt`` defers the magnitude
    to ``ParseUint``, which checks it INSIDE the scan and against ``uint64`` --
    not ``int64`` -- returning ``ErrRange`` the moment the accumulator would
    overflow, before it ever looks at the rest of the string. So the first
    disqualifying byte wins, and these two rows, one digit apart, refuse in
    opposite directions::

        "18446744073709551615x"   digits reach uint64 max, scan hits 'x' -> SYNTAX
        "18446744073709551616x"   the overflow fires first               -> RANGE

    Testing the whole string for well-formedness first ("is it all digits? then
    parse") gets that pair backwards, and backwards in the ACCEPTING direction:
    it answers the syntax refusal, which both callers read as the system actor,
    so an unreadable id silently becomes "basecamp" instead of failing.
    """
    # One optional ASCII sign. `+` IS accepted -- the `^-?\d+$` regex most ports
    # reach for is not this grammar.
    sign = text[:1]
    # An offset, not a slice: `text[1:]` copied the whole remaining id before the
    # scan even started. The scan itself stops at the first non-digit or at the
    # first overflow -- within ~20 significant digits -- so a long malformed id
    # cost a copy of itself for nothing. (Leading zeros carry no magnitude, so an
    # all-zero-padded id is still walked in full, exactly as Go walks it.)
    start = 1 if sign in ("+", "-") else 0
    # An empty digit run, with or without a sign, is a syntax error: "", "+", "-".
    if start == len(text):
        return Refusal.SYNTAX

    # `ParseUint`'s loop, byte for byte. `str.isdigit()` would be wrong here --
    # it is true for "৭" -- and so would `str.isdecimal()`; the test is on the
    # ASCII range itself.
    magnitude = 0
    for character in islice(text, start, None):
        if not ("0" <= character <= "9"):
            return Refusal.SYNTAX
        magnitude = magnitude * 10 + (ord(character) - ord("0"))
        # Checked every digit, not once at the end: Python would happily carry a
        # 200-digit accumulator, and the point is to refuse where Go refuses.
        if magnitude > _UINT64_MAX:
            return Refusal.RANGE

    # `ParseInt`'s own bound, applied to what `ParseUint` returned. The negative
    # side reaches one further than the positive: -9223372036854775808 is a
    # value, +9223372036854775808 is a range error.
    value = -magnitude if sign == "-" else magnitude
    if not (_INT64_MIN <= value <= _INT64_MAX):
        return Refusal.RANGE
    return value


def coerce_person_id(obj: MutableMapping[str, Any]) -> None:
    """Rewrite a Person-shaped object's string ``id`` in place, as Go's normalizer does.

    ``coercePersonID`` (``go/pkg/basecamp/normalize.go:40``), one outcome per
    ``ParseInt`` result:

    * a value becomes the ``int`` (``:58`` -- Go writes the CANONICAL digits
      back, so ``"+7"`` and ``"007"`` both land as ``7``, never held verbatim);
    * a syntax error is the non-numeric system-actor sentinel: id ``0`` with the
      original string preserved as ``system_label`` (``:66-67``);
    * a range error leaves the string UNTOUCHED (``:63``) so the reader refuses
      it. The readers that do refuse it are the ones with a Go struct behind
      them: ``services/_campfire_index._decoded_flexible_int64`` raises
      "overflows int64", and ``services/_merge_safe`` refuses a RANGE
      refusal as not a person id. Coercing it here instead -- which ``int()``
      was doing -- hands back a Python arbitrary-precision int for a value that
      does not fit the wire's ``int64``, and every reader downstream then
      believes it.

    An object whose ``id`` is already a number, or absent, is left alone -- which
    is also what makes this IDEMPOTENT, and :func:`normalize_person_ids` depends
    on that. A second call after a value or a label finds an ``int`` and returns;
    a second call after a range refusal re-reads the same string and refuses it
    again, for the same reason.
    """
    raw_id = obj.get("id")
    if not isinstance(raw_id, str):
        return
    outcome = parse_int64(raw_id)
    if outcome is Refusal.RANGE:
        return
    if outcome is Refusal.SYNTAX:
        obj["system_label"] = raw_id
        obj["id"] = 0
        return
    obj["id"] = outcome


def normalize_person_ids(obj: Any, *, embedded_people: bool = False) -> None:
    """Normalize every Person-shaped object in a decoded response body, in place.

    This stands where ``normalizeEmbeddedPeopleJSON`` stands
    (``go/pkg/basecamp/normalize.go:117-134``), so it owns BOTH of the passes
    that function runs, and it finds a person TWO ways:

    1. **By ``personable_type``** (``normalizePersonIds``, ``:16-29``) -- any
       object at any depth carrying that key, whatever its value.
    2. **By structural position** (``normalizeEmbeddedPersonIds``, ``:83-104``)
       -- the ``creator`` object and each element of the ``participants`` array,
       on any object at any depth, WHETHER OR NOT it carries a
       ``personable_type``.

    The second pass is not redundant, and Go's own comment says why (``:78-82``):
    embedded creator and participant people frequently omit ``personable_type``,
    so the first pass skips exactly the payloads the second exists to fix.

    Missing it is observable HERE, where in most ports it would not be. Every
    port with a runtime decoder behind the normalizer -- Kotlin's
    ``FlexibleLongSerializer``, Swift's ``FlexibleInt``, Rust's ``flexible_i64``
    -- converts the string at read time whichever pass did or did not touch it.
    The generated services hand back a plain ``dict``; the one decoder standing
    in for ``FlexibleInt64`` there is :func:`decode_person_id_sites`, which reads
    the id but writes no ``system_label``. So on a notification or gauge body
    only this pass gives an untagged system actor its label, as the reference's
    does.

    This pass does NOT close every string person id off the wire, and does not
    try to -- see "WHERE PASS 2 RUNS" below. Off Go's two surfaces a string id
    in ``creator``, ``assignees`` or a schedule's ``participants`` is decoder
    coverage rather than normalizer reach, and :func:`decode_person_id_sites`
    closes it at exactly the ``Person`` sites the schema declares.

    ONE walk here where Go runs two, and the two are equivalent because
    :func:`coerce_person_id` is idempotent and both passes apply it unchanged.
    A decoded body is a tree, so each node is reached once per pass; the set of
    coerced nodes is the union of "has ``personable_type``" and "is the
    ``creator`` or a ``participants`` element of its parent", which is the same
    set either way; and a node in both sets is coerced twice under both schemes,
    a no-op the second time. Only the order differs, and idempotence is what
    makes order not matter. One walk rather than two keeps this off a second
    full traversal of every response body.

    WHERE PASS 2 RUNS, which is a correction. Pass 1 runs on every body, as it
    always has: an object that declares ``personable_type`` IS the Person
    projection. Pass 2 runs only when ``embedded_people`` is true, which the
    service base sets for exactly the endpoints Go calls
    ``normalizeEmbeddedPeopleJSON`` from -- ``decodeGaugePayload``
    (``gauges.go:170``) and the notification decoders
    (``my_notifications.go:171, 281, 296``). See :data:`EMBEDDED_PEOPLE_PATHS`.

    It used to run on every body, on the reasoning that wider is safe when it is
    wider in the accepting direction. On these keys it is not. ``creator`` and
    ``participants`` also sit on ``UpcomingScheduleEntry``, where they hold
    ``UpcomingSchedulePerson`` -- a plain ``int64`` id in the reference
    (``client.gen.go:4194-4198``) -- so a string there is a decode error in Go,
    and this turned it into person 0 with a ``system_label``: the SYSTEM ACTOR,
    for a body the reference refuses outright. Accepting direction, identity
    field, which is the class this work exists to remove.

    Narrowing costs no coverage: schedules is not one of Go's two surfaces, and
    a string ``participants`` id on a ``Person`` site there is read by
    :func:`decode_person_id_sites`, as Go's typed decode reads it (SPEC.md
    section 10, "Person Ids Off the Wire").
    """
    if isinstance(obj, list):
        for item in obj:
            normalize_person_ids(item, embedded_people=embedded_people)
        return
    if not isinstance(obj, dict):
        return
    # Pass 1's test is the KEY'S PRESENCE, whatever its value (`:19`), not that
    # it is a string naming a type.
    if "personable_type" in obj:
        coerce_person_id(obj)
    # Pass 2's type assertions are Go's, mirrored: a `creator` that is not an
    # object and a `participants` that is not an array are not people, so they
    # are skipped rather than coerced (`:85-95`). `dict` is what Go's
    # `.(map[string]any)` accepts and a list is not.
    if embedded_people:
        creator = obj.get("creator")
        if isinstance(creator, dict):
            coerce_person_id(creator)
        participants = obj.get("participants")
        if isinstance(participants, list):
            for participant in participants:
                if isinstance(participant, dict):
                    coerce_person_id(participant)
    # Both of Go's passes then recurse through every child, so a creator nested
    # under a comment under an event is reached at whatever depth it sits
    # (`:22-24`, `:96-102`).
    for value in obj.values():
        if isinstance(value, (dict, list)):
            normalize_person_ids(value, embedded_people=embedded_people)


#: The endpoints the reference runs the POSITIONAL pass over, and only those.
#:
#: ``normalizeEmbeddedPeopleJSON`` is a function in the reference, not a layer:
#: it is called from ``decodeGaugePayload`` (every gauge and needle body) and
#: from the notification decoders, and from nowhere else. Matched on the request
#: path because Go's own boundary is the call site.
#:
#: Each pattern is anchored to the END of the URL's path, and only the path is
#: matched -- never the host, never the query. An unanchored pattern over the
#: whole URL would let a base URL whose own path contained ``/gauge_needles/``
#: switch the pass on for every request.
EMBEDDED_PEOPLE_PATHS: tuple[re.Pattern[str], ...] = (
    re.compile(r"/my/readings\.json\Z"),  # GetMyNotifications
    re.compile(r"/my/readings/bubble_ups\.json\Z"),  # GetBubbleUps
    re.compile(r"/gauge_needles/\d+\Z"),  # GetGaugeNeedle, UpdateGaugeNeedle
    re.compile(r"/projects/\d+/gauge/needles\.json\Z"),  # ListGaugeNeedles, CreateGaugeNeedle
    re.compile(r"/reports/gauges\.json\Z"),  # ListGauges
)


def embedded_people_url(url: str) -> bool:
    """Whether ``url`` is one of the reference's two normalization surfaces."""
    path = urlsplit(url).path
    return any(pattern.search(path) for pattern in EMBEDDED_PEOPLE_PATHS)


def decode_person_id_sites(
    body: Any,
    table: Mapping[str, Iterable[tuple[str, ...]]],
    operation: str | None,
    *,
    followed_page: bool = False,
    only_under: str | None = None,
) -> None:
    """Read each person id at ``operation``'s sites in ``table`` as Go's ``types.FlexibleInt64``, in place.

    Go types ``Person.id`` as ``types.FlexibleInt64``
    (``go/pkg/types/flexible_int64.go``) and every generated Go service decodes
    its body through ``Parse<Op>Response`` first, so an untagged ``{"id": "7"}``
    comes back as ``7`` wherever a ``Person`` sits. These dicts have no decoder,
    so this stands in for that one field, at exactly the sites the generator
    found by the ``x-go-type`` marker (``generated/services/_person_id_sites.py``)
    -- never by key name, which would reach the plain-int64 people
    (``UpcomingSchedulePerson`` and co.) where Go refuses a string.

    Runs AFTER :func:`normalize_person_ids`, on the same body the paths are
    relative to. Per id (``flexible_int64.go:28-63``):

    * a string is ``ParseInt``: a value becomes the ``int``; a syntax refusal
      becomes ``0`` with NO ``system_label`` (the decoder writes none, ``:47``);
      a range refusal fails the read (``:44``);
    * an ``int`` within int64 stays; any other JSON value -- float, out-of-range
      int, ``null``, bool, array, object -- fails the read (``:55-60``).

    A person that is not an object, or has no ``id`` key, and a container of
    the wrong shape, are left alone: Go zero-fills or refuses those as part of a
    whole-body typed decode this SDK does not do for any field. Idempotent,
    because every surviving id is an ``int``.

    FOLLOWED PAGES are not ``Parse<Op>Response``, and Go decodes less of them;
    the paginators pass what Go reads and flag it:

    * a bare-array list passes only the items the cap keeps -- ``followPagination``
      trims the raw items before its caller decodes any (``client.go:604-631``);
    * a wrapped listing passes ``only_under`` its items key, because Go reads a
      followed page as ``struct{ Events []json.RawMessage }`` and decodes each
      event, never the page's ``person`` (``timeline.go:444-457``);
    * ``followed_page`` on :data:`HAND_DECODED_FOLLOWED_PAGES` lets a ``null`` id
      through, below.
    """
    if operation is None:
        return
    null_passes = followed_page and operation in HAND_DECODED_FOLLOWED_PAGES
    for site in table.get(operation, ()):
        if only_under is None or site[:1] == (only_under,):
            _decode_site(body, site, 0, operation, null_passes)


#: Operations whose FOLLOWED pages Go decodes into hand-written types rather than
#: through the generated ``Person``: ``Gauge`` and ``GaugeNeedle``
#: (``gauges.go:236-241, 307-312``) and ``Notification``
#: (``my_notifications.go:285-305``), each after the positional normalizer. Their
#: ``Person.ID`` is a plain ``int64``, which reads JSON ``null`` as ``0`` with no
#: error where ``FlexibleInt64`` refuses it; every other refusal is the same. Go
#: SDK behaviour, not schema, so it is listed by hand. Page 1 of each still goes
#: through ``Parse<Op>Response`` and still refuses ``null``.
HAND_DECODED_FOLLOWED_PAGES: frozenset[str] = frozenset({"ListGauges", "ListGaugeNeedles", "GetBubbleUps"})


def _decode_site(node: Any, site: tuple[str, ...], index: int, operation: str | None, null_passes: bool) -> None:
    if index == len(site):
        # A passed `null` stays `null`, the same representation residual as an
        # absent id, rather than the `0` Go's plain int64 leaves.
        if isinstance(node, dict) and "id" in node and not (null_passes and node["id"] is None):
            node["id"] = _flexible_int64(node["id"], operation, site)
        return
    segment = site[index]
    if segment == "[]":
        children: Iterable[Any] = node if isinstance(node, list) else ()
    elif segment == "{}":
        children = node.values() if isinstance(node, dict) else ()
    else:
        children = (node[segment],) if isinstance(node, dict) and segment in node else ()
    for child in children:
        _decode_site(child, site, index + 1, operation, null_passes)


def _flexible_int64(raw: Any, operation: str | None, site: tuple[str, ...]) -> int:
    # `bool` is an `int` subclass in Python and a JSON boolean in Go: refused.
    if isinstance(raw, int) and not isinstance(raw, bool):
        if _INT64_MIN <= raw <= _INT64_MAX:
            return raw
    elif isinstance(raw, str):
        outcome = parse_int64(raw)
        if outcome is Refusal.SYNTAX:
            return 0
        if outcome is not Refusal.RANGE:
            return outcome
    where = ".".join(site) or "$"
    raise ApiError(
        f"{operation or 'response'}: person id at {where} is not an int64: {_truncate(json.dumps(raw, default=repr))}"
    )

File: typescript/src/person-id.ts
/**
 * Go's `strconv.ParseInt(s, 10, 64)` over a person id, written out.
 *
 * Three places in this SDK turn a person id that arrived as a JSON STRING into
 * a number, and they answer to the same Go line: the pre-decode normalizer in
 * `services/base.ts` stands where `coercePersonID` stands
 * (`go/pkg/basecamp/normalize.go:45`), and the read-side comparison in
 * `services/mentions.ts` and the merge-safe id-list guard in
 * `services/merge-safe.ts` stand where `FlexibleInt64` stands
 * (`go/pkg/types/flexible_int64.go:34`). All of them call
 * {@link scanPersonId}, so the rule cannot drift into rules that disagree about
 * the same string.
 *
 * Do not reach for `Number()`, `parseInt()`, or a `^-?\d+$` test in their place.
 * Each is wrong in a direction that matters and the three are wrong differently:
 * `Number()` accepts `"1e3"`, `"0x10"`, `"12.0"` and `" 7"`, all of which Go
 * refuses; `parseInt()` accepts a trailing tail, so `"7x"` reads 7; and the
 * regex refuses the leading `+` Go accepts, which is how `"+7"` became a system
 * actor. What none of them can do at all is the thing this file exists for —
 * tell Go's two refusals apart. See {@link PersonIdScan}.
 */

/**
 * What `strconv.ParseInt(s, 10, 64)` answered: a value, or WHICH WAY it refused.
 *
 * Keeping the two refusals apart is the whole point, because Go does two
 * different things with them and no single "is this a number?" predicate can
 * separate them. A SYNTAX error reads as `0` — Go's non-numeric sentinel, the
 * id of a `LocalPerson` (`"basecamp"`, `"campfire"`) — while a RANGE error is
 * raised and fails the read (`go/pkg/types/flexible_int64.go:43` against `:46`;
 * `go/pkg/basecamp/normalize.go:62-63` against `:66-67`). Which one you get
 * depends on where in the string the first disqualifying byte sits, so it
 * cannot be recovered after the fact from a yes/no answer.
 *
 * `value` is a `bigint` because Go's answer is an `int64` and a JS `number` is
 * not one. Rounding it here would erase, inside the scan, the very distinction
 * each caller has to make a decision about — see {@link personIdNumber}.
 */
export type PersonIdScan =
  | { readonly kind: "value"; readonly value: bigint }
  | { readonly kind: "syntax" }
  | { readonly kind: "range" };

const SYNTAX: PersonIdScan = { kind: "syntax" };
const RANGE: PersonIdScan = { kind: "range" };

const ZERO = 0x30;
const NINE = 0x39;
const PLUS = 0x2b;
const MINUS = 0x2d;

/** `ParseUint`'s bound: the magnitude is checked against u64, not i64. */
const U64_MAX = 18446744073709551615n;
/** `ParseInt`'s own bounds, applied to what `ParseUint` returned. */
const I64_MAX = 9223372036854775807n;
const I64_MIN_MAGNITUDE = 9223372036854775808n;

/**
 * `strconv.ParseInt(text, 10, 64)`, scan order included.
 *
 * The grammar is one optional ASCII sign, then one or more ASCII digits, and
 * nothing else:
 *
 * - A leading `+` is ACCEPTED, and a `^-?\d+$` regex is exactly the port defect
 *   this replaces — `"+7"` is 7, not a sentinel.
 * - No whitespace of any kind, because Go trims none: `" 7"`, `"7 "`, `"\t7"`
 *   are syntax errors and read `0`.
 * - Leading zeros carry no meaning at base 10, so `"007"` is 7 and `"010"` is
 *   TEN, not eight. (`ParseInt` only reads a prefix as a base when `base` is 0,
 *   which is why `"0x10"` is a syntax error here rather than 16.)
 * - `_` is a digit separator only at base 0, so `"1_0"` is a syntax error.
 * - ASCII `0`-`9` and nothing else: a fullwidth `７`, an Arabic-Indic `٧` and a
 *   Bengali `৭` are not digits, however `\p{Nd}`-aware a regex engine is. This
 *   scans UTF-16 units rather than code points on purpose — every unit of a
 *   non-ASCII character is outside `0x30..0x39`, including a lone surrogate, so
 *   the first one refuses exactly where Go's byte scan refuses.
 *
 * The subtlety that earns the hand-rolled loop: `ParseUint` checks the
 * magnitude INSIDE the scan and returns `ErrRange` the moment the accumulator
 * would overflow `uint64`, before it ever looks at the rest of the string. The
 * first disqualifying byte wins, and the bound it wins against is u64, not i64.
 * So `"18446744073709551616x"` is a RANGE error, which fails the read, while
 * `"18446744073709551615x"` — one digit smaller — is a SYNTAX error, which
 * reads `0`. Testing the whole string for well-formedness first and only then
 * checking the magnitude gets that pair backwards, which is the second defect
 * this replaces.
 *
 * This is NOT the rule the gid parser applies to the person id in
 * `gid://bc3/Person/<id>`. That one walks the bytes and refuses anything
 * outside `0..=9` BEFORE it parses (`go/pkg/basecamp/mentions.go:252-256`),
 * so it rejects the leading `+` this one accepts, and then refuses `id <= 0`.
 * Both rules live in this SDK: `personIdFromSGID` in `services/mentions.ts`
 * carries that digit walk, and it is CORRECT to, because the reference has that
 * shape at that site — a `+`-tolerant gid parser is the `+77` defect PR #886
 * closed. The same shape was wrong here only because the Go lines governing
 * these two sites have no pre-walk. Two sites, two rules, deliberately: do not
 * hoist either into the other, in either direction.
 */
export function scanPersonId(text: string): PersonIdScan {
  let index = 0;
  let negative = false;
  if (text.length > 0) {
    const sign = text.charCodeAt(0);
    if (sign === PLUS || sign === MINUS) {
      negative = sign === MINUS;
      index = 1;
    }
  }
  // An empty digit run is a syntax error: `""`, `"+"`, `"-"`.
  if (index === text.length) return SYNTAX;

  // `ParseUint`'s loop, unit for unit.
  let magnitude = 0n;
  for (; index < text.length; index++) {
    const code = text.charCodeAt(index);
    if (code < ZERO || code > NINE) return SYNTAX;
    magnitude = magnitude * 10n + BigInt(code - ZERO);
    // Go returns here, mid-scan, without reading what follows.
    if (magnitude > U64_MAX) return RANGE;
  }

  // `ParseInt`'s own bound, applied to what `ParseUint` returned. `-2^63` fits
  // and `+2^63` does not, so the two signs have different bounds.
  if (negative) {
    if (magnitude > I64_MIN_MAGNITUDE) return RANGE;
    // `-0n` is not a thing in BigInt, so `"-0"` yields `0n` and, through
    // `personIdNumber`, `+0` — where `Number("-0")` yielded `-0`, which
    // `Object.is` separates from the `0` Go reads.
    return { kind: "value", value: -magnitude };
  }
  if (magnitude > I64_MAX) return RANGE;
  return { kind: "value", value: magnitude };
}

const MIN_SAFE = BigInt(Number.MIN_SAFE_INTEGER);
const MAX_SAFE = BigInt(Number.MAX_SAFE_INTEGER);

/**
 * The id Go read, as a JS `number` — or `undefined` when a `number` cannot hold
 * it.
 *
 * This is the port's one unclosable gap, and the only judgment call in this
 * file. `ParseInt` read a REAL PERSON here: `"9007199254740993"` is an id, not
 * a refusal. A `number` cannot carry it — past `Number.MAX_SAFE_INTEGER` two
 * distinct int64s land on the same double, so `9007199254740993` would be
 * handed to the caller as `9007199254740992`, a different person — and the
 * SDK's whole surface types every id as `number`. `conformance/tests/
 * integer-precision.json` records the same constraint for the JSON-number path
 * in as many words: returning bigint would break the number-typed API surface.
 * So there is no value this can return, and TypeScript has no runtime decoder
 * downstream to perform Go's RANGE refusal on its behalf. Whatever is chosen, a
 * caller sees it. Three directions were available:
 *
 * - **Write `0` and a `system_label`** — what the normalizer did before this
 *   file existed. That is Go's NON-NUMERIC sentinel: the id of `LocalPerson`,
 *   `"basecamp"`, `"campfire"`. It takes a real person, hands back the system
 *   actor, and says nothing — in the exact shape a caller is meant to trust. A
 *   wrong id that looks right is the one outcome nothing downstream can defend
 *   against, and it is the defect this work exists to remove.
 * - **Throw** — refuses a response Go reads fine, and refuses ALL of it. One
 *   unrepresentable id in one embedded person would discard every other record
 *   in the body, including the 99 people that were perfectly readable, and it
 *   would do so from inside a normalizer that runs over every response. The id
 *   is safe; everything beside it pays.
 * - **Say "unreadable"** — what this returns, and what each site then resolves
 *   locally with the treatment the measured table already assigns to a RANGE
 *   refusal. The normalizer LEAVES THE STRING in place: nothing is rounded,
 *   nothing is invented, the digits survive verbatim for a caller that can hold
 *   them, and `typeof person.id === "string"` is a check a caller can actually
 *   make. The reader returns `undefined`, and `mentionMarkup` refuses to write
 *   a mention it cannot verify. Both are refusals of the ID, not of the
 *   response.
 *
 * `personIdValue` in `services/mentions.ts` already answered it this way before
 * the two sites shared a rule, which settles the consistency question in the
 * same direction: the normalizer joins the reader.
 *
 * This is a RESIDUAL DIVERGENCE and is recorded as one, not a fix. Over the
 * measured corpus, the 11 rows whose value falls outside ±(2^53 - 1) still
 * disagree with Go at both sites, and will until the SDK's ids become `bigint`.
 */
export function personIdNumber(value: bigint): number | undefined {
  if (value < MIN_SAFE || value > MAX_SAFE) return undefined;
  return Number(value);
}
